packages/@pollyjs/adapter-http/src/utils/transport-wrapper.js

```diff
@@ -0,0 +1,198 @@
+import http from 'http';
+import zlib from 'zlib';
+
+import { URL } from '@pollyjs/utils';
+
+const NATIVE_REQUEST = Symbol();
+const NATIVE_END = Symbol();
+
+const LISTENERS = Symbol();
+
+export default class TransportWrapper {
+  constructor(transport, { adapter }) {
+    this.adapter = adapter;
+    this.transport = transport;
+  }
+
+  patch() {
+    if (!this.transport[NATIVE_REQUEST]) {
+      this.transport[NATIVE_REQUEST] = this.transport.request;
+
+      this.transport.request = this.createRequestWrapper();
+    }
+  }
+
+  restore() {
+    if (this.transport[NATIVE_REQUEST]) {
+      this.transport.request = this.transport[NATIVE_REQUEST];
+
+      delete this.transport[NATIVE_REQUEST];
+    }
+  }
+
+  getHttpResponseData(res) {
+    return new Promise(resolve => {
+      let stream;
+
+      switch (res.headers['content-encoding']) {
+        case 'gzip':
+        case 'compress':
+        case 'deflate':
+          stream = res.pipe(zlib.createUnzip());
+          delete res.headers['content-encoding'];
+          break;
+        default:
+          stream = res;
+      }
+
+      const resBuffer = [];
+
+      stream.on('data', chunk => {
+        resBuffer.push(chunk);
+      });
+
+      stream.on('end', () => {
+        const data = Buffer.concat(resBuffer).toString('utf8');
```

Comment:
Do you see any issues with storing the data as `utf8`? Will there be an issues with replaying content with other encodings?

Reply:
I haven't tried with a binary content but it'll probably fail right now. Maybe we can store binary content as `base64` ?

Reply:
Im ok with storing binary data as `base64`, I do have a few questions/concerns:

1. Do we have a way of determining whether we can decode the buffer content in utf-8 or fallback to base64? Can we use response headers or maybe something like [is-utf8](https://www.npmjs.com/package/is-utf8)?
2. When we respond with the base64 content, do we need to provide the original encoding to the buffer? If so, should we persist the encoding in the recordings (I can help with that if necessary)?


@yasinuslu can you also add a test case for this as well?

Reply:
Follow up question:

Do we support for sending binary data in the request payload? We should add a test case for this as well. If not, I can help on the @pollyjs/core side to try to get this working.

Reply:
@offirgolan Of course, i'll add a test case for this. 

I don't have the answers you're looking for right now, i'll get back to you after some experimenting 😄

Reply:
I did some testing on this and i think it's best if we store every response body as `base64`. 
So in response to your questions if we store `base64`:
1. We won't need to determine which encoding to use.
2. We will respond with a buffer with original bytes because we respond with `Buffer.from(storedBody, 'base64')` we can construct same bytes whether they were binary or utf8.
3. We can probably support binary data in the request payload if we also store it as `base64`.

I successfully replayed both utf8 and binary data in my local, i'll add test cases for both binary request payload and binary responses. Maybe we can add a new field in the recordings that indicates if the stored body is utf8 encoded or base64.

Reply:
I did some investigation on my end as well. Looking at nock's implementation, we can do something very similar to determine how to handle both the request and response body:

https://github.com/nock/nock/blob/master/lib/recorder.js#L49-L97

Reply:
Thanks, that seems like a nice library. I'll start implementing something similar than.